File: tf2/tf_python_ws/src/static_broadcaster/static_broadcaster/turtle_tf2_listener.py
# Copyright 2021 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Đoạn mã nguồn được viết với kiểu giao tiếp service kết hợp giao tiếp qua topic
import math

from geometry_msgs.msg import Twist

import rclpy
from rclpy.node import Node

from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
# Các đoạn mã nguồn trước là gửi đi phép biển đổi
from tf2_ros.transform_listener import TransformListener # Lắng nghe biến đổi

from turtlesim.srv import Spawn # Loại service


class FrameListener(Node):

    def __init__(self):
        super().__init__('turtle_tf2_frame_listener')

        # Declare and acquire `target_frame` parameter
        # Tạo ra param có tên là target_fram với giá trị khởi tạo là turtle1
        self.target_frame = self.declare_parameter(
            'target_frame', 'turtle1').get_parameter_value().string_value

        self.tf_buffer = Buffer() # Bộ nhớ lưu chữ các transform (TF)
        self.tf_listener = TransformListener(self.tf_buffer, self)

        # Create a client to spawn a turtle
        # Tạo client yêu cầu sinh ra con rùa
        self.spawner = self.create_client(Spawn, 'spawn')
        # Boolean values to store the information
        # if the service for spawning turtle is available
        # Đặt cờ hiệu để biết nếu service sinh con rùa đã sẵn sang chạy
        self.turtle_spawning_service_ready = False
        # if the turtle was successfully spawned
        # Cờ hiệu để biết con rùa đã sinh ra chưa ?
        self.turtle_spawned = False

        # Create turtle2 velocity publisher
        self.publisher = self.create_publisher(Twist, 'turtle2/cmd_vel', 1)

        # Call on_timer function every second
        self.timer = self.create_timer(1.0, self.on_timer)

    def on_timer(self):
        # Store frame names in variables that will be used to
        # compute transformations
        from_frame_rel = self.target_frame # Frame cần theo (mục tiêu)
        to_frame_rel = 'turtle2' # Frame rùa đang ở

        if self.turtle_spawning_service_ready: # Kiểm tra service đã sẵn sàng chưa ? 
            if self.turtle_spawned: # Kiểm tra con rùa 2 đã được tạo chưa ? 
                # Look up for the transformation between target_frame and turtle2 frames
                # and send velocity commands for turtle2 to reach target_frame
                try:
                    t = self.tf_buffer.lookup_transform(
                        to_frame_rel,
                        from_frame_rel,
                        rclpy.time.Time())
                    # Lấy transform của turtle1 ->target_frame (ở đây là lấy vector)
                # Nếu chưa có transfrom thì log lỗi và thoát hàm on_timer  
                except TransformException as ex:
                    self.get_logger().info(
                        f'Could not transform {to_frame_rel} to {from_frame_rel}: {ex}')
                    return
                # Quá trình điều khiển con rùa

                msg = Twist() # Tạo lệnh điều khiển vận tốc
                scale_rotation_rate = 1.0 # Hệ số khuyếch đại ( tốc độ phản ứng góc khi target thay đổi hướng đi)
                msg.angular.z = scale_rotation_rate * math.atan2(
                    t.transform.translation.y,
                    t.transform.translation.x
                # Góc cân quay là atan2(y,x), đảm bảo con rùa sẽ luôn hướng về target    
                )
                # Vận tốc tối đa
                scale_forward_speed = 0.5
                msg.linear.x = scale_forward_speed * math.sqrt(
                    t.transform.translation.x ** 2 +
                    t.transform.translation.y ** 2
                #Công thức: distance = sqrt(x^2 +y^2)
                )
                # Xuất bản thông cho tới /turtl2/cmd_vel để điều khiển
                self.publisher.publish(msg)
            # Kiểm tra service nếu chưa spawn 
            else:
                if self.result.done():
                    self.get_logger().info(
                        f'Successfully spawned {self.result.result().name}')
                    self.turtle_spawned = True
                else:
                    self.get_logger().info('Spawn is not finished')
        # Nếu CHƯA gửi request spawn
        else:
            if self.spawner.service_is_ready():
                # Initialize request with turtle name and coordinates
                # Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
                request = Spawn.Request()
                request.name = 'turtle2'
                request.x = 4.0 # Tọa độ x ngay khi được Spawn
                request.y = 2.0 # Tọa độ y ngay khi được Spawn
                request.theta = 0.0 # Góc hướng

                # Call request
                self.result = self.spawner.call_async(request)
                self.turtle_spawning_service_ready = True
            else:
                # Check if the service is ready
                self.get_logger().info('Service is not ready')


def main():
    rclpy.init()
    node = FrameListener()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    rclpy.shutdown()

File: tf2/tf_cpp_ws/src/learning_tf2_cpp/src/turtle_tf2_listener.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp" 


#include "tf2/exceptions.h"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "turtlesim/srv/spawn.hpp"

class FrameListenser : public rclcpp::Node{
public:
    FrameListenser(): Node("turtle_tf2_frame_listener"),
        turtle_spawn_service_ready_(false),
        turtle_spawned_(false)
    {
        // Chọn mục tiêu target
        target_frame_ = this->declare_parameter<std::string>("target_frame","turtle1");
        /*
            tf_buffer : là đối tượng sẽ lưu trữ các phép biến đổi của tf2
        */
        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        // Tạo listener
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
        // Tạo ra một client yêu cầu sinh ra con rùa 2
        spawner_ = this->create_client<turtlesim::srv::Spawn>("spawn");
        // Tạo ra publisher để xuất bản lệnh điều khiển con rùa 2
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle2/cmd_vel",1);

        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            // //Cách 1: dùng bind
            // std::bind(&FrameListenser::on_timer,this)
            // Cách 2: dùng lambda
            [this](){
                this->on_timer();
            }
        );
    }

private:
    void on_timer(){
        // Lấy tên mục tiêu
        std::string fromFrameRel = target_frame_.c_str();
        // Lấy tên đối tượng cần điều khiển
        std::string toFrameRel = "turtle2";

        if(turtle_spawn_service_ready_){
            if(turtle_spawned_){    
                geometry_msgs::msg::TransformStamped t;
                try{
                    t = tf_buffer_ ->lookupTransform(
                        toFrameRel,
                        this->get_clock()->now(),
                        fromFrameRel,
                        //tf2::TimePointZero //Ban đầu
                        this->get_clock()->now() - rclcpp::Duration(5,0), // Thử đổi sang đoạn này
                        "world",
                        std::chrono::microseconds(100)
                    );
                }
                catch(const tf2::TransformException &ex){
                    RCLCPP_INFO(this->get_logger(),
                        "could not transform %s to %s : %s",
                        fromFrameRel.c_str(),toFrameRel.c_str(),
                        ex.what()
                    );
                    return ;
                }
                
                geometry_msgs::msg::Twist msg;
                // Điều khiển góc quay

                static const double scaleRoationRate = 1.0;
                msg.angular.z = scaleRoationRate*atan2(
                    t.transform.translation.y,
                    t.transform.translation.x
                );

                double dx = t.transform.translation.x;
                double dy = t.transform.translation.y;
                static const double scaleForwardSpeed  = 0.5;
                double distance = hypot(dx, dy);
                if (distance < 0.05) {
                    msg.linear.x  = 0.0;
                    msg.angular.z = 0.0;
                }
                else {
                    msg.angular.z = 1.0 * atan2(dy, dx);
                    msg.linear.x  = scaleForwardSpeed * distance;
                }
                
                publisher_ ->publish(msg);
            }
            else{
                RCLCPP_INFO(this->get_logger(), "Successfully spawned");
                turtle_spawned_ = true;
            }
        }
        else{
            if (spawner_->service_is_ready()){
                // Khởi tạo yêu cầu sinh ra con rùa cùng với tọa độ thiết lập
                auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
                request->x = 4.0;
                request->y = 3.0;
                request->theta = 0.0;
                request->name = "turtle2";

                // Gọi request
                using ServiceResponseFuture = rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture;
                auto response_recieved_callback = [this](ServiceResponseFuture future){
                    auto result = future.get();
                    if(strcmp(result->name.c_str(),"turtle2")==0){
                        turtle_spawn_service_ready_ = true;
                    }
                    else{
                        RCLCPP_INFO(this->get_logger(),
                            "Service callback rasult mismatch");
                    }
                };
                auto result = spawner_->async_send_request(request,response_recieved_callback);
            }
            else{
                RCLCPP_INFO(this->get_logger(),
                    "Service is not ready");
            }
        }
    }
    bool turtle_spawn_service_ready_;
    bool turtle_spawned_ ;
    std::string target_frame_;
    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawner_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};
int main(int argc, char **argv){
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<FrameListenser>());
    rclcpp::shutdown();
    return 0;
}